src/model_repository_manager.cc

```diff
@@ -671,155 +653,237 @@ ModelRepositoryManager::LoadUnloadModels(
     const std::unordered_map<
         std::string, std::vector<const InferenceParameter*>>& models,
     const ActionType type, const bool unload_dependents,
-    bool* all_models_polled)
+    bool* all_models_polled,
+    std::shared_ptr<ConditionVariable>* conflict_retry_cv)
 {
-  auto status = Status::Success;
-  *all_models_polled = true;
-  // Update ModelInfo related to file system accordingly
-  std::set<ModelIdentifier> added, deleted, modified, unmodified;
+  // They are here to hold information for loading/unloading models later, and
+  // also hold after loading/unloading information to be written back to this
+  // object later.
+  ModelInfoMap infos;
+  std::unordered_map<std::string, std::set<ModelIdentifier>> global_map;
+  DependencyGraph dependency_graph;
+
+  // deleted, deleted_dependents are here for async unload models later.
+  // affected_models is here for unlocking loading/unloading models later.
+  std::set<ModelIdentifier> deleted, deleted_dependents, affected_models;
+
   {
-    if (type == ActionType::UNLOAD) {
-      for (const auto& model : models) {
-        // Within the class, model is referred by ModelIdentifier, so it needs
-        // to check global map to find identifier assoicated with the name.
-        auto git = global_map_.find(model.first);
-        if (git != global_map_.end()) {
-          for (const auto& mid : git->second) {
-            deleted.insert(mid);
-          }
-        }
+    std::lock_guard<std::mutex> lock(mu_);
+
+    // Prepare model infos related to file system accordingly, do not write yet
+    // until committed to load/unload the models.
+    // Also poll the added, deleted, modified, unmodified models.
+    infos = infos_;
+    std::set<ModelIdentifier> added, modified, unmodified;
+    RETURN_IF_ERROR(LoadUnloadModelsPoll(
```

Comment:
Give it different name, i.e. `PollModel()`, and should avoid keep passing the `type` because `UNLOAD` really doesn't need to perform the poll at all, just resovle the different type here:
```
switch (type) {
    case LOAD:
         PollModels(...);
    case UNLOAD:
         ListDeleted(...); // can be different name or just process the result in place.
}
```

Reply:
Update commit: Rename LoadUnloadModelsPoll to PollModels